tests/validate/test_facades.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2023 Canonical Ltd.
+# Licensed under the Apache V2, see LICENCE file for details.
+
+import importlib
+import re
+import warnings
+from pathlib import Path
+from types import ModuleType
+from typing import Dict, List, Optional, Set, Tuple, TypedDict, cast
+
+import pytest
+
+from juju.client import connection
+
+
+class Versions(TypedDict, total=True):
+    versions: List[int]
+
+
+ClientFacades = Dict[str, Versions]
+
+
+@pytest.fixture
+def project_root(request: pytest.FixtureRequest):
+    return request.config.rootpath
+
+
+class TestFacades:
+    def test_client_facades(self, project_root: Path) -> None:
+        good_facades = self.generate_client_facades(project_root)
+        client_facades = cast(ClientFacades, connection.client_facades)
+
+        errors: List[Tuple[str, Optional[List[int]], Optional[List[int]]]] = []
+        all_names = sorted(set(connection.client_facades).union(good_facades))
+        for name in all_names:
+            expected = self._versions_from_facades(good_facades, name)
+            actual = self._versions_from_facades(client_facades, name)
+            if expected != actual:
+                errors.append((name, expected, actual))
+
+        if errors:
+            print('The following errors were found in connection.client_facades:')
```

Comment:
Let's refactor `connection.client_facades` and remove the intermediate "versions" key.

Then the test can be reduced to:

```suggestion
        generated: Dict[str, List[int]] = scan_the_client_NN_files()
        
        assert client_facades == generated
```
        
pytest will neatly show the difference is there is any.

Reply:
Happy to remove all this if the pytest output is actually useful. I probably didn't scroll up enough, but I added it because the test failure message for `good_facades == client_facades` didn't look useful to me, as it heavily truncated both dicts.

Reply:
> Let's refactor connection.client_facades and remove the intermediate "versions" key.

We *could* do this but I'd rather leave it in till we finish the schema work as I'd also have to refactor my various `client_facades` generation scripts (from `api/facadeversions.go`, from `_clientN.py`, from schema files). I mean it's a pretty simple refactor, but still worth a separate PR I think.

Reply:
There you go #1155